Public/Src/Cache/DistributedCache.Host/Service/Internal/DistributedContentStoreFactory.cs

```diff
@@ -320,43 +321,66 @@ private async Task ApplySecretSettingsForLlsAsync(RedisContentLocationStoreConfi
                 value => eventStoreConfiguration.MaxEventProcessingConcurrency = value);
         }
 
-        private AzureBlobStorageCredentials[] GetStorageCredentials(Dictionary<string, string> settings, StringBuilder errorBuilder)
+        private AzureBlobStorageCredentials[] GetStorageCredentials(Dictionary<string, Secret> secrets, StringBuilder errorBuilder)
         {
             var storageSecretNames = GetAzureStorageSecretNames(errorBuilder);
-            if (storageSecretNames == null)
+            // This would have failed earlier otherwise
+            Contract.Requires(storageSecretNames != null);
+
+            var credentials = new List<AzureBlobStorageCredentials>();
+            foreach (var secretName in storageSecretNames)
             {
-                return null;
+                var secret = GetRequiredSecret(secrets, secretName);
+
+                if (!_distributedSettings.AzureBlobStorageUseSasTokens && secret is PlainTextSecret plainTextSecret)
+                {
+                    credentials.Add(new AzureBlobStorageCredentials(plainTextSecret.Secret));
+                }
+                else if (_distributedSettings.AzureBlobStorageUseSasTokens && secret is UpdatingSasToken updatingSasToken)
+                {
+                    var storageCredentials = new StorageCredentials(sasToken: updatingSasToken.Token.Token);
+                    updatingSasToken.TokenUpdated += (token, sasToken) =>
+                    {
+                        storageCredentials.UpdateSASToken(sasToken.Token);
+                    };
+
+                    // The account name should never actually be updated, so its OK to take it from the initial token
+                    credentials.Add(new AzureBlobStorageCredentials(storageCredentials, updatingSasToken.Token.StorageAccount));
+                }
+                else
+                {
+                    throw new NotSupportedException($"Unexpected secret kind retrieved for requested secret {secretName}");

```

Comment:
I think you can actually assert that if AzureBlobStorageUseSasTokens is true then secret **must be** of type UpdatingSasToken.

I think if this is not true, then we have a bug in the code.

And we can use assertions to distinguish bugs from expected exceptions.